br/pkg/lightning/backend/external/util.go

```diff
@@ -186,3 +187,50 @@ func MockExternalEngine(
 	}
 	return GetAllFileNames(ctx, storage, "/mock-test")
 }
+
+// EndpointTp is the type of Endpoint.Key.
+type EndpointTp int
+
+const (
+	// ExclusiveEnd represents "..., Endpoint.Key)".
+	ExclusiveEnd EndpointTp = iota
+	// InclusiveStart represents "[Endpoint.Key, ...".
+	InclusiveStart
+	// InclusiveEnd represents "..., Endpoint.Key]".
+	InclusiveEnd
+)
+
+// Endpoint represents an endpoint of an interval which can be used by GetMaxOverlapping.
+type Endpoint struct {
+	Key    []byte
+	Tp     EndpointTp
+	Weight uint64 // all EndpointTp use positive weight
+}
+
+// GetMaxOverlapping returns the maximum overlapping weight treating given
+// `points` as endpoints of intervals. `points` are not required to be sorted,
+// and will be sorted in-place in this function.
+func GetMaxOverlapping(points []Endpoint) int {
+	slices.SortFunc(points, func(i, j Endpoint) int {
+		if bytes.Compare(i.Key, j.Key) != 0 {
+			return bytes.Compare(i.Key, j.Key)
```

Comment:
Can we avoid comparing twice?

Reply:
https://github.com/pingcap/tidb/pull/46614/commits/3377a054f4fa40ed1a4404255b7cd0b271b1274b